src/dev/i18n/extract_html_messages.js

```diff
@@ -0,0 +1,142 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { jsdom } from 'jsdom';
+import { parse } from '@babel/parser';
+import { isDirectiveLiteral, isObjectExpression, isStringLiteral } from '@babel/types';
+
+import { isPropertyWithKey, escapeLineBreak, traverseNodes } from './utils';
+import { DEFAULT_MESSAGE_KEY, CONTEXT_KEY } from './constants';
+
+/**
+ * Find all substrings of "{{ any text }}" pattern
+ */
+const ANGULAR_EXPRESSION_REGEX = /\{\{+([\s\S]*?)\}\}+/g;
+
+const I18N_FILTER_MARKER = '| i18n: ';
+
+/**
+ * Extract default message from an angular filter expression argument
+ * @param {string} expression JavaScript code containing a filter object
+ * @returns {string} Default message
+ */
+function parseFilterObjectExpression(expression) {
+  // parse an object expression instead of block statement
+  const nodes = parse(`+${expression}`).program.body;
+
+  for (const node of traverseNodes(nodes)) {
+    if (!isObjectExpression(node)) {
+      continue;
+    }
+
+    let message;
+    let context;
+
+    for (const property of node.properties) {
+      if (isPropertyWithKey(property, DEFAULT_MESSAGE_KEY)) {
+        if (!isStringLiteral(property.value)) {
+          throw new Error('defaultMessage value should be a string literal.');
+        }
+
+        message = escapeLineBreak(property.value.value);
+      } else if (isPropertyWithKey(property, CONTEXT_KEY)) {
+        if (!isStringLiteral(property.value)) {
+          throw new Error('context value should be a string literal.');
+        }
+
+        context = property.value.value;
+      }
+    }
+
+    return { message, context };
+  }
+}
+
+function parseIdExpression(expression) {
+  for (const node of traverseNodes(parse(expression).program.directives)) {
+    if (isDirectiveLiteral(node)) {
+      return node.value;
+    }
+  }
+}
+
+function trimCurlyBraces(string) {
+  return string.slice(2, -2).trim();
+}
+
+function* getFilterMessages(htmlContent) {
+  const expressions = (htmlContent.match(ANGULAR_EXPRESSION_REGEX) || [])
+    .filter(expression => expression.includes(I18N_FILTER_MARKER))
+    .map(trimCurlyBraces);
+
+  for (const expression of expressions) {
+    const filterStart = expression.indexOf(I18N_FILTER_MARKER);
+    const idExpression = expression.slice(0, filterStart).trim();
+    const filterObjectExpression = expression.slice(filterStart + I18N_FILTER_MARKER.length).trim();
+
+    if (!filterObjectExpression || !idExpression) {
+      throw new Error(`Cannot parse i18n filter expression: {{ ${expression} }}`);
+    }
+
+    const messageId = parseIdExpression(idExpression);
+
+    if (!messageId) {
+      throw new Error('Empty "id" value in angular filter expression is not allowed.');
+    }
+
+    try {
+      const { message, context } = parseFilterObjectExpression(filterObjectExpression);
```

Comment:
__note:__ destructuring will likely fail badly if `parseFilterObjectExpression` returns null or undefined, you probably want to have `parseFilterObjectExpression(filterObjectExpression) || {}`.